lib/mp4/transmuxer.js

```diff
@@ -291,7 +309,190 @@ VideoSegmentStream = function(track) {
     // Continue with the flush process now
     this.trigger('done');
   };
+
+  // create the default sample
+  // see ISO/IEC 14496-12:2012, section 8.6.4.3
+  this.createDefaultSample_ = function() {
+    return {
+      size: 0,
+      flags: {
+        isLeading: 0,
+        dependsOn: 1,
+        isDependedOn: 0,
+        hasRedundancy: 0,
+        degradationPriority: 0
+      }
+    };
+  };
+
+  // search an array of nal units to see if it qualifies as an i-frame (actually, IDR)
+  this.frameIsIFrame_ = function(frame) {
+    var
+      i,
+      currentNal;
+
+    for (i = 0; i < frame.length; i++) {
+      currentNal = frame[i];
+
+      if (currentNal.nalUnitType === 'slice_layer_without_partitioning_rbsp_idr') {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  this.filterLeadingNonIFrames_ = function(frames) {
+    var
+      i,
+      currentFrame,
+      initialPts = frames[0][0].pts;
+
+    for (i = 0; i < frames.length; i++) {
+      currentFrame = frames[i];
+      if (this.frameIsIFrame_(currentFrame)) {
+        currentFrame[0].pts = initialPts;
+        break;
+      }
+    }
+    return frames.slice(i);
+  };
+
+  // convert an array of nal units into an array of frames with each frame being
+  // composed of the nal units that make up that frame
+  this.groupNalsIntoFrames_ = function(nalUnits) {
+    var
+      i,
+      currentNal,
+      currentFrame,
+      frames = [];
+
+    for (i = 0; i < nalUnits.length; i++) {
+      currentNal = nalUnits[i];
+
+      if (currentNal.nalUnitType === 'access_unit_delimiter_rbsp') {
+        if (currentFrame && currentFrame.length) {
+          frames.push(currentFrame);
```

Comment:
If we made each element of `frames` an object instead, we could note whether it was a keyframe or not here without significant additional complexity.